File: seed/problems/002_valid-brackets/solutions/solution.cpp
#include <iostream>
#include <stack>
#include <string>

using namespace std;

int main() {
    string s;
    getline(cin, s);

    stack<char> stk;
    for (char c : s) {
        if (c == '(' || c == '[' || c == '{') {
            stk.push(c);
        } else if (c == ')' || c == ']' || c == '}') {
            if (stk.empty()) {
                cout << "No" << endl;
                return 0;
            }
            char top = stk.top();
            stk.pop();
            if ((c == ')' && top != '(') ||
                (c == ']' && top != '[') ||
                (c == '}' && top != '{')) {
                cout << "No" << endl;
                return 0;
            }
        }
    }

    if (stk.empty()) {
        cout << "Yes" << endl;
    } else {
        cout << "No" << endl;
    }

    return 0;
}

File: seed/problems/002_valid-brackets/solutions/solution.py
import sys


def main():
    s = sys.stdin.readline().strip()

    stack = []
    matching = {')': '(', ']': '[', '}': '{'}

    for c in s:
        if c in '([{':
            stack.append(c)
        elif c in ')]}':
            if not stack or stack[-1] != matching[c]:
                print("No")
                return
            stack.pop()

    print("Yes" if not stack else "No")


if __name__ == "__main__":
    main()
